File: cpp/remove_nth_node_from_end_of_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        ListNode dummy(-1);
        dummy.next = head;
        ListNode *p, *p_n;
        p = &dummy;
        p_n = &dummy;
        for (int i=0; i<n; ++i)
            p_n = p_n->next;
        while (p_n->next)
        {
            p_n = p_n->next;
            p = p->next;
        }
        ListNode *tmp = p->next;
        p->next = p->next->next;
        delete tmp;
        return dummy.next;
    }
};

File: python/remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        node_array = []
        tmp = head
        length = 0
        while tmp != None:
            node_array.append(tmp)
            tmp = tmp.next
            length += 1
        if length == n:
            return head.next
        else:
            node_array[length-n-1].next = node_array[length-n].next
            return head